File: generated-code/codecraft/rust/src/model/build_properties.rs
use super::*;

#[derive(Clone, Debug)]
pub struct BuildProperties {
    pub options: Vec<EntityType>,
    pub init_health: Option<i32>,
}

impl trans::Trans for BuildProperties {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.options.write_to(writer)?;
        self.init_health.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let options: Vec<EntityType> = trans::Trans::read_from(reader)?;
        let init_health: Option<i32> = trans::Trans::read_from(reader)?;
        Ok(Self {
            options,
            init_health,
        })
    }
}

File: generated-code/codecraft/scala/src/main/scala/model/BuildProperties.scala
package model

import util.StreamUtil

case class BuildProperties(options: Seq[model.EntityType], initHealth: Option[Int]) {
    def writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, options.length)
        options.foreach { value =>
            value.writeTo(stream)
        }
        initHealth match {
            case None => StreamUtil.writeBoolean(stream, false)
            case Some(value) => {
                StreamUtil.writeBoolean(stream, true)
                StreamUtil.writeInt(stream, value)
            }
        }
    }

    override def toString(): String = {
        var stringBuilder = new StringBuilder("BuildProperties { ")
        stringBuilder.append("options: ")
        stringBuilder.append(options)
        stringBuilder.append(", ")
        stringBuilder.append("initHealth: ")
        stringBuilder.append(initHealth)
        stringBuilder.append(" }")
        stringBuilder.toString()
    }
}

object BuildProperties {
    def readFrom(stream: java.io.InputStream): BuildProperties = BuildProperties(
        (0 until StreamUtil.readInt(stream)).map { _ =>
            model.EntityType.readFrom(stream)
        },
        if (StreamUtil.readBoolean(stream)) Some(
            StreamUtil.readInt(stream)
        ) else None
    )
}